File: src/main/scala/com/github/rockjam/itgm7/scala/Closures.scala
/*
 * Copyright 2016 Nikolay Tatarinov
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.github.rockjam.itgm7.scala

import com.github.rockjam.itgm7.java

object Closures {

  val foo = "first foo"
  val bar = "first bar"
  val baz = "first baz"

  def makePrinter(foo: String, bar: String): () => String = {
    () =>
      {
        val foo = "third foo"
        foo + " " + bar + " " + baz
      }
  }

  val closure = makePrinter("second foo", "second bar")

  closure()
//  > third foo second bar first baz

  UseClosure.cl() == closure() //should be true

  val minValue = 5

  val minFilter: Int => Boolean = i => i > minValue

  Vector(9, 2, 3, 4, 1, 3, 2, 9, 4, 0, 6, 0, 2, 7, 9, 3, 6, 8, 1, 1, 6, 5, 5, 8, 9, 8, 7, 6, 1, 3) filter minFilter


}

object UseClosure {
  val foo = "other foo"
  val bar = "other bar"
  val baz = "other baz"

  val cl = Closures.closure
}

File: src/main/java/com/github/rockjam/itgm7/java/Closures.java
package com.github.rockjam.itgm7.java;

import java.util.ArrayList;
import java.util.List;

public class Closures {

    public List<Integer> closureFilter() {

Integer minValue = 5;
Filter<Integer> filter = new MinFilter(minValue);

List<Integer> ints = new ArrayList<>();
ints.add(1);
ints.add(2);
ints.add(4);
ints.add(6);
ints.add(8);
ints.add(12);
ints.add(8);

List<Integer> filtered = new ArrayList<>();
for (Integer i : ints) {
    if(filter.filter(i)) filtered.add(i);
}
return filtered;
    }

    public class MinFilter implements Filter<Integer> {
        private Integer minValue;

        public MinFilter(Integer minValue) {
            this.minValue = minValue;
        }

        @Override
        public boolean filter(Integer i) {
            return i > minValue;
        }
    }


}
